### ISSUE TITLE
[INSTALL]: gsi-ncdiag/1.1.2, bufr/11.7.0, met/9.1.3, metplus/3.1.1

### ISSUE THREAD

**Issue by DavidHuber-NOAA** [JCSDA/spack-stack#841]
### Package name

gsi-ncdiag, bufr, met, and metplus

### Package version/tag

1.1.2, 11.7.0, 9.1.3, and 3.1.1, respectively

### Build options

BUILD_TYPE=`Release` for all

### Installation timeframe

I would like these to be added into the 1.5.0 installations so that the GSI, UFS_Utils, etc, can use the same spack-stack as the UFS once PR ufs-community/ufs-weather-model#1920 is merged.  If we could start with Hera and Orion, that would be great, with a preference for Hera.

### Other information

Work was performed to fix the interface between the GSI and bufr/12.0.0, but there are optimization (run time) issues still to be resolved, so we would like to continue using 11.7.0 for the time being until those can be resolved.

gsi-ncdiag/1.1.2 is an upgrade and fixes a memory over-allocation in the GSI.

MET and METplus are used for the global verification system (verif-global).

@AlexanderRichert-NOAA installed a test environment for the global workflow with these versions and with spack-stack-1.4.1.  This was used to compile the GSI, GSI-Utils, GSI-Monitor, GFS-Utils, and UFS_Utils packages and then run the global workflow.  Tests were successful at C96/C48 and C384/C192 resolutions (see NOAA-EMC/global-workflow#1868 for more details).

-----------
**Install round 1**
- [x] Hera: /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core
- [x] Orion: /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core
- [x] Jet: /mnt/lfs4/HFIP/hfv3gfs/role.epic/spack-stack/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core
- [x] Gaea C5: /lustre/f2/dev/wpo/role.epic/contrib/spack-stack/c5/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core
- [x] Hercules: /work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core
- [x] S4: /data/prod/jedi/spack-stack/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core

**Checked for/added gsi-ncdiag@1.1.2, added wrf-io@1.2.0 and sfcio@1.4.1**
How to add wrf-io@1.2.0 / sfcio@1.4.1 ? They are already installed in the upstream environment, therefore:
```
spack install --add --verbose wrf-io 2>&1 | tee envs/gsi-addon/log.install.wrf-io-1.2.0
spack install --add --verbose sfcio@1.4.1 2>&1 | tee envs/gsi-addon/log.install.sfcio-1.4.1
spack module lmod refresh --upstream-modules
```
- [x] Hera: /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core
- [x] Orion: /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core
- [x] Jet: /mnt/lfs4/HFIP/hfv3gfs/role.epic/spack-stack/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core
- [x] Gaea C5: /lustre/f2/dev/wpo/role.epic/contrib/spack-stack/c5/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core
- [x] Hercules: /work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core
- [x] S4: /data/prod/jedi/spack-stack/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core

**Comment by AlexanderRichert-NOAA** [comment]
@DavidHuber-NOAA would it be better from Global Workflow's perspective if https://github.com/ufs-community/ufs-weather-model/pull/1920 used these chained environments instead? In other words, if the 1.5.0-based environment we add bufr 11.7.0 and gsi-ncdiag 1.1.2 into is called "gw-env", then would it make sense (or make any difference) for the MODULEPATH setting in UFS to point to /path/to/spack-stack-1.5.0/gw-env/install/modulefiles/Core as opposed to /path/to/spack-stack-1.5.0/unified-env/install/modulefiles/Core ?

**Comment by DavidHuber-NOAA** [comment]
Hmm, it would be nice if we could all get along and use the same environment.  However, I don't think that it is necessary in this case.  So I guess I would say yes, it would be better from the global workflow's perspective, but I wouldn't want to add significant work to PR1920.

On a different note, I recall now that there are two additional libraries we discussed adding: met/9.1.3 and metplus/3.1.1.  Could that also be added in to this environment for the verif-global package?  I believe the dependencies for these packages could remain the same as for met/10.x metplus/4.x (i.e. you could build with bufr/11.7.0 or bufr/12.0.0, either way would be fine).  I will amend my title/answers above.

**Comment by AlexanderRichert-NOAA** [comment]
Yep no problem

**Comment by climbfuji** [comment]
Can we at some point in the near future try to converge to the latest versions of these packages so that we only need to install one version of each? Happy to set up a meeting to discuss this. Thanks!

**Comment by DavidHuber-NOAA** [comment]
@climbfuji There is work ongoing to fix some overly long runtimes with bufr/12 within the GSI.  I suspect an upgrade to BUFR will be coming along soon that the GSI can use.

As for met and metplus, I am afraid that the system these are needed for ([verif-global](https://github.com/NOAA-EMC/EMC_verif-global)) will not go through the upgrade process to implement newer versions of those libraries.  There is a replacement on the horizon ([EVS](https://github.com/NOAA-EMC/Evs)), but there isn't a timeline for that at the moment.  That said, verif-global can be run offline and stick with the current hpc-stack installs, so the installation of met/metplus isn't critical.  So if it is a significant effort/difficulty, we can move forward without installing these older versions in spack-stack.

**Comment by DavidHuber-NOAA** [comment]
FYI @aerorahul

**Comment by climbfuji** [comment]
Can EPIC or EMC coordinate the install of those packages? Should we do this based on spack-stack 1.5.1 instead of 1.5.0, maybe in a separate environment? @AlexanderRichert-NOAA @ulmononian ?

**Comment by DavidHuber-NOAA** [comment]
I think adding these to spack-stack/1.5.1 instead of 1.5.0 would be fine, especially since gsi-ncdiag/1.1.2 is already there.  Ideally, I would like them in the same environment that the UFS will use (at least bufr/11.7.0), but it isn't a problem if they end up in a new environment.

**Comment by AlexanderRichert-NOAA** [comment]
I tested this out on spack-stack-1.5.1 on Acorn: /lfs/h1/emc/nceplibs/noscrub/spack-stack/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core

For some reason the met configure step failed until I added the python lib path to LD_LIBRARY_PATH in the build environment...

@DavidHuber-NOAA can you confirm that the versions look right, and if so I can install on Hera and Orion?

**Comment by DavidHuber-NOAA** [comment]
@AlexanderRichert-NOAA Thanks for the test install.  I don't yet have access to Acorn (only Dogwood and Cactus).  I have requested it this morning which hopefully will be approved soon.  I will let you know when I am able to test this.

**Comment by DavidHuber-NOAA** [comment]
@AlexanderRichert-NOAA @aerorahul and I sat down and went through the versions -- everything looks good!  Thanks!

**Comment by AlexanderRichert-NOAA** [comment]
@climbfuji here are the steps I used (I haven't run the whole thing start to finish so caveat emptor etc.): [gsiaddon.txt](https://github.com/JCSDA/spack-stack/files/13245864/gsiaddon.txt)

**Comment by AlexanderRichert-NOAA** [comment]
Hera, Orion, Jet, C5, and Hercules are done

**Comment by climbfuji** [comment]
@DavidHuber-NOAA I followed Alex's instructions on S4. Please start with
```
module use /data/prod/jedi/spack-stack/spack-stack-1.5.0/envs/gsi-addon/install/modulefiles/Core
```
and the rest should be as usual. Can you check if I got it done right please? Thanks!

**Comment by climbfuji** [comment]
I just realized I did this for 1.5.0 instead of 1.5.1 on S4, sorry. Will go back and do it again ...

**Comment by climbfuji** [comment]
Ok, next try: `/data/prod/jedi/spack-stack/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/Core`

**Comment by DavidHuber-NOAA** [comment]
@climbfuji I tried out a few utilities.  ncdump and wgrib2 were successful.  prepbufr2nc failed with a segmentation fault, though I noticed NOAA-EMC/nceplibs-bufr#528 addressed a build problem with that binary, so this may be more or less expected.

Next, I checked that all necessary libraries were available for the GSI.  It appears that one is missing: wrf-io/1.2.0.  Also, the gsi-ncdiag version is 1.1.1 rather than 1.1.2.

I checked this against Hera.  The wrf-io installation is also missing there and is also not part of the 1.5.1/unified-env environment.  The gsi-ncdiag/1.1.2 library is available in both the gsi-addon and unified-env environments.

Could you please upgrade gsi-ncdiag on S4 to 1.1.2 and install wrf-io/1.2.0 under the gsi-addon environment?

@AlexanderRichert-NOAA could I please get wrf-io/1.2.0 installed in the gsi-addon environment on all systems?  Apologies for not checking this when you installed the test environment on Acorn.

**Comment by climbfuji** [comment]
@DavidHuber-NOAA Done on S4. @AlexanderRichert-NOAA I updated the checklist in https://github.com/JCSDA/spack-stack/issues/841#issue-1954946478 at the bottom for the wrf-io step.

@DavidHuber-NOAA, if this GSI addon environment is something we need to support for longer, should we consider having a template for it in configs/templates, and then install it either standalone or using unified-env as upstream on selected platforms each time? We have at the moment in configs/templates:
```
drwxr-xr-x 2 jedipara domain users 11776 Oct 31 21:12 empty
drwxr-xr-x 2 jedipara domain users 11776 Oct 31 21:12 gfs-v16.2
drwxr-xr-x 2 jedipara domain users 11776 Oct 31 21:12 skylab-dev
drwxr-xr-x 2 jedipara domain users 11776 Oct 31 21:12 ufs-srw-dev
drwxr-xr-x 2 jedipara domain users 11776 Oct 31 21:12 ufs-srw-public-v2
drwxr-xr-x 2 jedipara domain users 11776 Oct 31 21:12 ufs-utils
drwxr-xr-x 2 jedipara domain users 11776 Oct 31 21:12 ufs-weather-model
drwxr-xr-x 2 jedipara domain users 11776 Oct 31 21:12 ufs-weather-model-static
drwxr-xr-x 2 jedipara domain users 11776 Oct 31 21:12 unified-dev
```

**Comment by DavidHuber-NOAA** [comment]
@climbfuji Thanks for the build.

@climbfuji @AlexanderRichert-NOAA It seems I missed one other library: sfcio/1.4.1.  I am going to go through all of the libraries needed for the global-workflow and all of its subrepositories and just verify there is nothing else.

I think that a template would be a good idea.  I think that the BUFR situation can be resolved fairly soon, but MET/METplus will probably take some time.  Having the unified-env as the upstream environment makes a lot of sense to me.

**Comment by DavidHuber-NOAA** [comment]
@climbfuji @AlexanderRichert-NOAA I verified that all other modules are available on Hera and S4.  So on S4, could I please get sfcio/1.4.1 installed?  On all other systems, could I get sfcio/1.4.1 and wrf-io/1.2.0 installed?

**Comment by climbfuji** [comment]
@DavidHuber-NOAA That's done on S4 (and once again I updated the description at the top of this issue). Can I ask you to work on the template for this special environment?

**Comment by DavidHuber-NOAA** [comment]
@climbfuji Sure, I will create an issue for it now.

**Comment by DavidHuber-NOAA** [comment]
@climbfuji The GSI successfully compiled with gsi-addon on S4.  However, I noticed during the build process that most of the module files are pointing to spack-stack/1.5.0 builds, not 1.5.1.  For instance, in `/data/prod/jedi/spack-stack/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/intel/2021.5.0/sigio/2.3.2.lua`, we have
```
prepend_path("LD_LIBRARY_PATH", "/data/prod/jedi/spack-stack/spack-stack-1.5.0/envs/unified-env/install/intel/2021.5.0/sigio-2.3.2-g2swh3o/lib", ":")
```

So it seems that unified-env/1.5.0 was used as the upstream rather than 1.5.1.  I think that I am OK with this for S4, but it is worth noting.

**Comment by climbfuji** [comment]
I don't

> @climbfuji The GSI successfully compiled with gsi-addon on S4. However, I noticed during the build process that most of the module files are pointing to spack-stack/1.5.0 builds, not 1.5.1. For instance, in `/data/prod/jedi/spack-stack/spack-stack-1.5.1/envs/gsi-addon/install/modulefiles/intel/2021.5.0/sigio/2.3.2.lua`, we have
> 
> ```
> prepend_path("LD_LIBRARY_PATH", "/data/prod/jedi/spack-stack/spack-stack-1.5.0/envs/unified-env/install/intel/2021.5.0/sigio-2.3.2-g2swh3o/lib", ":")
> ```
> 
> So it seems that unified-env/1.5.0 was used as the upstream rather than 1.5.1. I think that I am OK with this for S4, but it is worth noting.

That must have been a stupid mistake on my end. Should I redo it or are you ok with it?

**Comment by DavidHuber-NOAA** [comment]
@climbfuji I am OK with it.  We can call S4 done at this point.

**Comment by climbfuji** [comment]
> @climbfuji I am OK with it. We can call S4 done at this point.

Ok, thanks. Sorry for that mishap.

**Comment by AlexanderRichert-NOAA** [comment]
@DavidHuber-NOAA I'll add sfcio and wrf_io shortly. The fms version will soon change in the next day or two as well (2023.02.01->2023.03) but I don't think that will affect you.

**Comment by AlexanderRichert-NOAA** [comment]
sfcio and wrf_io are added on all the listed systems

**Comment by DavidHuber-NOAA** [comment]
Thanks @AlexanderRichert-NOAA, I try out test builds of the GSI on Hera, Orion, and Jet today.

**Comment by DavidHuber-NOAA** [comment]
I built the GSI on Hera (both GNU and Intel), Orion, and Jet successfully.  I also verified the versions were as expected on Gaea and Hercules.

**Comment by DavidHuber-NOAA** [comment]
I believe everything is as it should be and that this can be closed.  Thank you very much @AlexanderRichert-NOAA and @climbfuji!
